cmd/jujud/agent/caasunitinit.go

```diff
@@ -0,0 +1,201 @@
+
+package agent
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/juju/cmd"
+	"github.com/juju/errors"
+	"github.com/juju/gnuflag"
+	"github.com/juju/loggo"
+	"gopkg.in/juju/names.v3"
+
+	"github.com/juju/juju/agent/tools"
+	"github.com/juju/juju/caas"
+	jujucmd "github.com/juju/juju/cmd"
+	cmdutil "github.com/juju/juju/cmd/jujud/util"
+	"github.com/juju/juju/juju/sockets"
+	"github.com/juju/juju/worker/uniter"
+	"github.com/juju/juju/worker/uniter/runner/jujuc"
+)
+
+// CAASUnitInitCommand represents a jujud bootstrap command.
+type CAASUnitInitCommand struct {
+	cmd.CommandBase
+
+	Wait bool
+	Send bool
+	Args struct {
+		Unit               string `json:"unit"`
+		OperatorFile       string `json:"operator-file"`
+		OperatorCACertFile string `json:"operator-ca-cert-file"`
+		CharmDir           string `json:"charm-dir"`
+	}
+
+	copyFunc      func(string, string) error
+	symlinkFunc   func(string, string) error
+	removeAllFunc func(string) error
+	mkdirAllFunc  func(string, os.FileMode) error
+}
+
+// NewCAASUnitInitCommand returns a new CAASUnitInitCommand that has been initialized.
+func NewCAASUnitInitCommand() *CAASUnitInitCommand {
+	return &CAASUnitInitCommand{
+		copyFunc:      copy,
+		symlinkFunc:   os.Symlink,
+		removeAllFunc: os.RemoveAll,
+		mkdirAllFunc:  os.MkdirAll,
+	}
+}
+
+// Info returns a description of the command.
+func (c *CAASUnitInitCommand) Info() *cmd.Info {
+	return jujucmd.Info(&cmd.Info{
+		Name:    "caas-unit-init",
+		Purpose: "initialize caas unit filesystem",
+	})
+}
+
+func (c *CAASUnitInitCommand) SetFlags(f *gnuflag.FlagSet) {
+	f.BoolVar(&c.Wait, "wait", false, "wait for args init via socket")
+	f.BoolVar(&c.Send, "send", false, "send args for init via socket")
+	f.StringVar(&c.Args.Unit, "unit", "", "unit name")
+	f.StringVar(&c.Args.OperatorFile, "operator-file", "", "operator client info file")
+	f.StringVar(&c.Args.OperatorCACertFile, "operator-ca-cert-file", "", "ca cert for operator")
+	f.StringVar(&c.Args.CharmDir, "charm-dir", "", "directory containing the charm")
+}
+
+func (c *CAASUnitInitCommand) Init(args []string) error {
+	if c.Wait && c.Send {
+		return errors.New("only one of --wait or --send can be specified")
+	}
+	return nil
+}
+
+func (c *CAASUnitInitCommand) Run(ctx *cmd.Context) (errOut error) {
+	sock := sockets.Socket{
+		Network: "unix",
+		Address: "@jujud-caas-unit-init.socket",
+	}
+	if c.Wait {
+		l, err := sockets.Listen(sock)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		defer l.Close()
+		conn, err := l.Accept()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		defer conn.Close()
+		err = json.NewDecoder(conn).Decode(&c.Args)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		w := loggo.NewSimpleWriter(conn, loggo.DefaultFormatter)
+		err = loggo.RegisterWriter("socket", w)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	} else if c.Send {
```

Comment:
can we move this block first since it returns
then the wait stuff can go beneath it with the rest of the logic below

Reply:
if resolving without doing, could we add a comment to explain why